### PR TITLE
Order

### PR DESCRIPTION
**Pull request by ihorromaniuk**

### CODE REVIEW

File: src/main/java/core/basesyntax/bookstore/model/OrderItem.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+@Getter
+@Setter
+@Entity
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/core/basesyntax/bookstore/model/OrderItem.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/core/basesyntax/bookstore/model/Order.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    @Column(nullable = false)
+    private BigDecimal total;
+
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
+
+    @Column(nullable = false)
+    private String shippingAddress;
+
+    @OneToMany(fetch = FetchType.EAGER, mappedBy = "order")
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/Order.java]:
```suggestion
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/core/basesyntax/bookstore/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    private Set<OrderItem> orderItems;
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/Order.java]:
```suggestion
    private Set<OrderItem> orderItems = new HashSet<>();
```

File: src/main/java/core/basesyntax/bookstore/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/model/Order.java]:
```suggestion
```

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        List<OrderItemDto> orderItemDtos = orderItems
+                .stream()
+                .map(orderMapper::toOrderItemDto)
+                .toList();
+        return new PageImpl<>(orderItemDtos, pageable, orderItems.getTotalElements());
+    }
+
+    @Override
+    public OrderItemDto getOrderItemById(Long orderId, Long orderItemId, User user) {
+        if (orderRepository.existsByIdAndUser(orderId, user)) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java]: you can use `OrderItem item = itemRepository.findByIdAndOrderIdAndUserId(orderItemId, orderId, userId)`

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    @Override
+    public OrderDto getOrderById(Long orderId, User user) {
+        return orderMapper.toDto(orderRepository.findByIdAndUser(orderId, user).orElseThrow(() ->
+                new EntityNotFoundException("Can't find order by id. Id: " + orderId)));
+    }
+
+    @Override
+    public Page<OrderItemDto> getOrderItems(Long orderId, Pageable pageable, User user) {
+        Order order = orderRepository.findByIdAndUser(orderId, user).orElseThrow(() ->
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java:84]: Page is redundant here

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    @Transactional
+    public OrderDto createOrder(CreateOrderRequestDto requestDto, User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.getByUser(user);
+        if (shoppingCart.getCartItems().isEmpty()) {
+            throw new ItemsNotFoundException("No items found in shopping cart to create order");
+        }
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(orderMapper::orderItemFromCartItem)
+                .collect(Collectors.toSet());
+        cartItemRepository.removeAllByShoppingCart(shoppingCart);
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java]: You can create method in ShoppingCart to remove cartItems
```
public void clearCart() {
        cartItems.clear();
    }
```

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                .orderDate(LocalDateTime.now())
+                .status(Order.Status.CREATED)
+                .orderItems(orderItems)
+                .total(total)
+                .user(user)
+                .build();
+        orderRepository.save(order);
+
+        orderItems.forEach(orderItem -> orderItem.setOrder(order));
+        orderItemRepository.saveAll(orderItems);
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java]: if you have cascade all this saving is redundant (check it )

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java (hunk 7 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java:84]: Order can contain a lot of items so I don't see how it's redundant

File: src/main/java/core/basesyntax/bookstore/model/Order.java (hunk 5 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/model/Order.java:28]: I need it for builder

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java (hunk 6 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java:100]: order item has no info about user

File: src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java (hunk 9 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/service/impl/OrderServiceImpl.java:74]: can't - should set order first

File: src/main/java/core/basesyntax/bookstore/model/Order.java (hunk 3 above)
Comment (ihorromaniuk) [reply src/main/java/core/basesyntax/bookstore/model/Order.java]: Can't use cascade save - have to set orderItem.order before saving it

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**